File: cli-proxy-logger/src/outbound.js
// Outbound proxy (出站代理) — opt-in, zero-dependency.
//
// Routes the proxy's UPSTREAM connections through an external HTTP/HTTPS or
// SOCKS5 proxy (common on locked-down intranets that only allow egress via a
// corporate proxy). Configure with a single URL, e.g.
//   UPSTREAM_PROXY=http://user:pass@proxy.corp:8080
//   UPSTREAM_PROXY=socks5://10.0.0.1:1080
// When unset, connections are made directly (default, unchanged behavior).
//
// Implementation: we expose a custom http(s).Agent whose createConnection()
// establishes the tunnel itself — an HTTP CONNECT for http/https proxies, or a
// SOCKS5 handshake for socks proxies — then, for https upstreams, performs the
// TLS handshake on top of the tunneled socket. No third-party packages.

import http from 'node:http';
import https from 'node:https';
import net from 'node:net';
import tls from 'node:tls';
import { createKernel, isKernelScheme } from './kernel/index.js';

// Parse a proxy URL into a normalized descriptor, or null when nothing/invalid.
// Supported protocols: http, https, socks/socks5/socks5h (all SOCKS5).
export function parseProxyUrl(raw) {
  if (!raw || typeof raw !== 'string' || raw.trim() === '') return null;
  let u;
  try {
    u = new URL(raw.trim());
  } catch {
    return null;
  }
  const scheme = u.protocol.replace(':', '').toLowerCase();
  let kind;
  if (scheme === 'http') kind = 'http';
  else if (scheme === 'https') kind = 'https';
  else if (scheme === 'socks' || scheme === 'socks5' || scheme === 'socks5h') kind = 'socks5';
  else return null;
  const defaultPort = kind === 'http' ? 80 : kind === 'https' ? 443 : 1080;
  return {
    kind,
    hostname: u.hostname,
    port: u.port ? Number.parseInt(u.port, 10) : defaultPort,
    username: u.username ? decodeURIComponent(u.username) : '',
    password: u.password ? decodeURIComponent(u.password) : '',
  };
}

// --- HTTP CONNECT tunnel -----------------------------------------------------
function httpConnect(proxy, host, port, cb) {
  const mod = proxy.kind === 'https' ? https : http;
  const headers = {};
  if (proxy.username || proxy.password) {
    const token = Buffer.from(`${proxy.username}:${proxy.password}`).toString('base64');
    headers['Proxy-Authorization'] = `Basic ${token}`;
  }
  const req = mod.request({
    host: proxy.hostname,
    port: proxy.port,
    method: 'CONNECT',
    path: `${host}:${port}`,
    headers,
    // The CONNECT request to the proxy itself; for an https proxy we don't
    // verify its cert chain against the target name (it's the proxy, not target).
    rejectUnauthorized: false,
  });
  req.once('connect', (res, socket) => {
    if (res.statusCode !== 200) {
      socket.destroy();
      cb(new Error(`proxy CONNECT failed: ${res.statusCode}`));
      return;
    }
    cb(null, socket);
  });
  req.once('error', cb);
  req.end();
}

// --- SOCKS5 handshake --------------------------------------------------------
// A tiny incremental reader so we can consume exact byte counts from the socket.
function makeReader(socket) {
  let buf = Buffer.alloc(0);
  const waiters = [];
  const pump = () => {
    while (waiters.length && buf.length >= waiters[0].n) {
      const w = waiters.shift();
      const out = buf.subarray(0, w.n);
      buf = buf.subarray(w.n);
      w.resolve(out);
    }
  };
  socket.on('data', (d) => {
    buf = Buffer.concat([buf, d]);
    pump();
  });
  return (n) => new Promise((resolve, reject) => {
    waiters.push({ n, resolve, reject });
    pump();
  });
}

function socks5Connect(proxy, host, port, cb) {
  const socket = net.connect(proxy.port, proxy.hostname);
  let done = false;
  const fail = (err) => {
    if (done) return;
    done = true;
    socket.destroy();
    cb(err);
  };
  socket.once('error', fail);
  socket.once('connect', async () => {
    try {
      const read = makeReader(socket);
      const useAuth = !!(proxy.username || proxy.password);
      // Greeting: VER=5, methods = [no-auth] (+ user/pass when creds given).
      socket.write(Buffer.from(useAuth ? [0x05, 0x02, 0x00, 0x02] : [0x05, 0x01, 0x00]));
      const greet = await read(2);
      if (greet[0] !== 0x05) throw new Error('socks5: bad version in greeting');
      const method = greet[1];
      if (method === 0xff) throw new Error('socks5: no acceptable auth method');
      if (method === 0x02) {
        const u = Buffer.from(proxy.username, 'utf8');
        const p = Buffer.from(proxy.password, 'utf8');
        socket.write(Buffer.concat([Buffer.from([0x01, u.length]), u, Buffer.from([p.length]), p]));
        const authRes = await read(2);
        if (authRes[1] !== 0x00) throw new Error('socks5: authentication failed');
      } else if (method !== 0x00) {
        throw new Error(`socks5: unsupported auth method ${method}`);
      }
      // CONNECT command with a domain-name target (ATYP=3) — lets the proxy
      // resolve DNS (socks5h semantics), which is what intranet setups want.
      const hostBuf = Buffer.from(host, 'utf8');
      const reqBuf = Buffer.concat([
        Buffer.from([0x05, 0x01, 0x00, 0x03, hostBuf.length]),
        hostBuf,
        Buffer.from([(port >> 8) & 0xff, port & 0xff]),
      ]);
      socket.write(reqBuf);
      const head = await read(4);
      if (head[1] !== 0x00) throw new Error(`socks5: connect failed (reply ${head[1]})`);
      const atyp = head[3];
      if (atyp === 0x01) await read(4);
      else if (atyp === 0x04) await read(16);
      else if (atyp === 0x03) {
        const len = (await read(1))[0];
        await read(len);
      } else throw new Error('socks5: bad ATYP in reply');
      await read(2); // BND.PORT
      done = true;
      socket.removeListener('error', fail);
      cb(null, socket);
    } catch (err) {
      fail(err);
    }
  });
}

function rawConnect(proxy, host, port, cb) {
  if (proxy.kind === 'socks5') socks5Connect(proxy, host, port, cb);
  else httpConnect(proxy, host, port, cb);
}

// Build the http/https tunnel agents for a (possibly deferred) proxy descriptor.
//   getProxy()  -> the current proxy descriptor (null until ready)
//   whenReady   -> resolves once getProxy() returns a usable descriptor
function buildAgents(getProxy, whenReady) {
  // Resolve readiness then perform the tunnel connect.
  function connectVia(host, port, cb) {
    whenReady.then(() => {
      const proxy = getProxy();
      if (!proxy) { cb(new Error('outbound proxy unavailable')); return; }
      rawConnect(proxy, host, port, cb);
    }, cb);
  }

  // http upstream: hand back the tunneled raw socket as-is.
  class HttpTunnelAgent extends http.Agent {
    createConnection(options, callback) {
      const host = options.host || options.hostname;
      const port = Number(options.port) || 80;
      connectVia(host, port, callback);
    }
  }
  // https upstream: wrap the tunneled socket in TLS for the target host.
  class HttpsTunnelAgent extends https.Agent {
    createConnection(options, callback) {
      const host = options.host || options.hostname;
      const port = Number(options.port) || 443;
      connectVia(host, port, (err, socket) => {
        if (err) {
          callback(err);
          return;
        }
        const tlsSock = tls.connect({
          socket,
          servername: options.servername || host,
          rejectUnauthorized: options.rejectUnauthorized !== false,
        });
        tlsSock.once('error', callback);
        tlsSock.once('secureConnect', () => {
          tlsSock.removeListener('error', callback);
          callback(null, tlsSock);
        });
      });
    }
  }

  const httpAgent = new HttpTunnelAgent({ keepAlive: false });
  const httpsAgent = new HttpsTunnelAgent({ keepAlive: false });
  return (upstreamUrl) => {
    const isHttps = String(upstreamUrl.protocol || upstreamUrl).includes('https');
    return isHttps ? httpsAgent : httpAgent;
  };
}

// Build the outbound proxy. Returns null when nothing is configured.
//
//   raw         a proxy/share URL: http/https/socks5 (direct, no kernel) OR an
//               advanced share link (vmess/vless/trojan/ss/hysteria2/tuic)
//               which is routed through a local xray/sing-box kernel.
//   kernelOpts  { kernel, configPath, xrayBin, singboxBin, socksPort } — when
//               configPath is set, the kernel is used regardless of `raw`.
export function createOutbound(raw, kernelOpts = {}) {
  const useKernel = isKernelScheme(raw) || !!(kernelOpts && kernelOpts.configPath);

  // --- direct path (today's behavior, unchanged) ---------------------------
  if (!useKernel) {
    const proxy = parseProxyUrl(raw);
    if (!proxy) return null;
    return {
      proxy,
      kernel: null,
      whenReady: Promise.resolve(),
      describe: `${proxy.kind}://${proxy.hostname}:${proxy.port}${proxy.username ? ' (auth)' : ''}`,
      agentFor: buildAgents(() => proxy, Promise.resolve()),
      stop() {},
    };
  }

  // --- kernel path (advanced protocols via xray/sing-box) ------------------
  const kernel = createKernel({
    link: isKernelScheme(raw) ? raw : undefined,
    configPath: kernelOpts.configPath,
    kernel: kernelOpts.kernel,
    xrayBin: kernelOpts.xrayBin,
    singboxBin: kernelOpts.singboxBin,
    socksPort: kernelOpts.socksPort,
  });
  let proxy = null; // socks5 -> local kernel inbound, set once ready
  const whenReady = kernel.start().then(({ socksPort }) => {
    proxy = { kind: 'socks5', hostname: '127.0.0.1', port: socksPort, username: '', password: '' };
  });
  // Surface a clear error rather than an unhandled rejection if the kernel dies.
  whenReady.catch((err) => console.error(`[kernel] startup failed: ${err.message}`));

  return {
    get proxy() { return proxy; },
    kernel,
    whenReady,
    describe: `kernel ${kernel.describe()}`,
    agentFor: buildAgents(() => proxy, whenReady),
    stop() { kernel.stop(); },
  };
}

File: cli-proxy-logger-py/cli_proxy_logger/outbound.py
"""Outbound proxy (出站代理) -- opt-in, zero-dependency.

Routes the proxy's UPSTREAM connections through an external HTTP/HTTPS or SOCKS5
proxy (common on locked-down intranets that only allow egress via a corporate
proxy). Configure with a single URL, e.g.::

    UPSTREAM_PROXY=http://user:pass@proxy.corp:8080
    UPSTREAM_PROXY=socks5://10.0.0.1:1080

When unset, connections are made directly (default, unchanged behavior).

Implementation: we establish the tunnel ourselves -- an HTTP CONNECT for
http/https proxies, or a SOCKS5 handshake for socks proxies -- then hand the
tunneled socket to a small ``http.client`` connection subclass (wrapping it in
TLS first for https upstreams). No third-party packages.
"""

import atexit
import base64
import http.client
import socket
import ssl
from urllib.parse import urlparse

from .kernel import create_kernel, is_kernel_scheme


def parse_proxy_url(raw):
    """Parse a proxy URL into a normalized dict, or None when nothing/invalid.
    Supported schemes: http, https, socks/socks5/socks5h (all SOCKS5)."""
    if not isinstance(raw, str) or raw.strip() == "":
        return None
    try:
        u = urlparse(raw.strip())
    except (ValueError, TypeError):
        return None
    scheme = (u.scheme or "").lower()
    if scheme == "http":
        kind = "http"
    elif scheme == "https":
        kind = "https"
    elif scheme in ("socks", "socks5", "socks5h"):
        kind = "socks5"
    else:
        return None
    if not u.hostname:
        return None
    default_port = 80 if kind == "http" else 443 if kind == "https" else 1080
    return {
        "kind": kind,
        "hostname": u.hostname,
        "port": u.port or default_port,
        "username": u.username or "",
        "password": u.password or "",
    }


def _http_connect_tunnel(proxy, host, port, timeout):
    """Open a TCP socket to the target via an HTTP CONNECT proxy."""
    sock = socket.create_connection((proxy["hostname"], proxy["port"]), timeout)
    try:
        if proxy["kind"] == "https":
            ctx = ssl._create_unverified_context()
            sock = ctx.wrap_socket(sock, server_hostname=proxy["hostname"])
        lines = [f"CONNECT {host}:{port} HTTP/1.1", f"Host: {host}:{port}"]
        if proxy["username"] or proxy["password"]:
            token = base64.b64encode(f"{proxy['username']}:{proxy['password']}".encode()).decode()
            lines.append(f"Proxy-Authorization: Basic {token}")
        sock.sendall(("\r\n".join(lines) + "\r\n\r\n").encode())
        buf = b""
        while b"\r\n\r\n" not in buf:
            chunk = sock.recv(4096)
            if not chunk:
                raise OSError("proxy closed connection during CONNECT")
            buf += chunk
        status_line = buf.split(b"\r\n", 1)[0].decode("latin1")
        parts = status_line.split(" ", 2)
        if len(parts) < 2 or parts[1] != "200":
            raise OSError(f"proxy CONNECT failed: {status_line}")
        return sock
    except Exception:
        sock.close()
        raise


def _socks5_tunnel(proxy, host, port, timeout):
    """Open a TCP socket to the target via a SOCKS5 proxy (with optional auth)."""
    sock = socket.create_connection((proxy["hostname"], proxy["port"]), timeout)
    try:
        use_auth = bool(proxy["username"] or proxy["password"])
        sock.sendall(bytes([0x05, 0x02, 0x00, 0x02]) if use_auth else bytes([0x05, 0x01, 0x00]))
        greet = _recv_exact(sock, 2)
        if greet[0] != 0x05:
            raise OSError("socks5: bad version in greeting")
        method = greet[1]
        if method == 0xFF:
            raise OSError("socks5: no acceptable auth method")
        if method == 0x02:
            u = proxy["username"].encode()
            p = proxy["password"].encode()
            sock.sendall(bytes([0x01, len(u)]) + u + bytes([len(p)]) + p)
            auth_res = _recv_exact(sock, 2)
            if auth_res[1] != 0x00:
                raise OSError("socks5: authentication failed")
        elif method != 0x00:
            raise OSError(f"socks5: unsupported auth method {method}")
        host_b = host.encode()
        req = bytes([0x05, 0x01, 0x00, 0x03, len(host_b)]) + host_b + bytes([(port >> 8) & 0xFF, port & 0xFF])
        sock.sendall(req)
        head = _recv_exact(sock, 4)
        if head[1] != 0x00:
            raise OSError(f"socks5: connect failed (reply {head[1]})")
        atyp = head[3]
        if atyp == 0x01:
            _recv_exact(sock, 4)
        elif atyp == 0x04:
            _recv_exact(sock, 16)
        elif atyp == 0x03:
            ln = _recv_exact(sock, 1)[0]
            _recv_exact(sock, ln)
        else:
            raise OSError("socks5: bad ATYP in reply")
        _recv_exact(sock, 2)  # BND.PORT
        return sock
    except Exception:
        sock.close()
        raise


def _recv_exact(sock, n):
    buf = b""
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise OSError("socks5: connection closed early")
        buf += chunk
    return buf


def _raw_connect(proxy, host, port, timeout):
    if proxy["kind"] == "socks5":
        return _socks5_tunnel(proxy, host, port, timeout)
    return _http_connect_tunnel(proxy, host, port, timeout)


def _build_helper(proxy, describe, kernel=None):
    """Wrap a normalized socks5/http proxy dict into the outbound helper object.
    ``proxy`` may be a callable returning the dict (kernel sets it once ready)."""
    resolve = proxy if callable(proxy) else (lambda: proxy)

    class _TunnelHTTPConnection(http.client.HTTPConnection):
        def connect(self):
            self.sock = _raw_connect(resolve(), self.host, self.port, self.timeout)

    class _TunnelHTTPSConnection(http.client.HTTPSConnection):
        def connect(self):
            raw_sock = _raw_connect(resolve(), self.host, self.port, self.timeout)
            ctx = self._context or ssl.create_default_context()
            self.sock = ctx.wrap_socket(raw_sock, server_hostname=self.host)

    def connection(scheme, host, port, timeout=600):
        if scheme == "https":
            return _TunnelHTTPSConnection(host, port, timeout=timeout)
        return _TunnelHTTPConnection(host, port, timeout=timeout)

    return {
        "proxy": resolve,
        "describe": describe,
        "connection": connection,
        "kernel": kernel,
        "stop": (kernel.stop if kernel else (lambda: None)),
    }


def create_outbound(raw, kernel_opts=None):
    """Build an outbound helper bound to a proxy URL, or None when unset/invalid.

    The returned object exposes ``connection(scheme, host, port, timeout)`` which
    yields an ``http.client`` connection whose socket is tunneled through the
    proxy (and TLS-wrapped for https upstreams).

    ``raw`` may be a plain http/https/socks5 proxy URL (used directly) or an
    advanced share link (vmess/vless/trojan/ss/hysteria2/tuic) which is routed
    through a local xray/sing-box kernel. ``kernel_opts`` (dict) may carry
    ``kernel`` (auto/xray/sing-box), ``configPath``, ``xrayBin``, ``singboxBin``,
    ``socksPort``."""
    kernel_opts = kernel_opts or {}
    use_kernel = is_kernel_scheme(raw) or bool(kernel_opts.get("configPath"))

    # --- direct path (today's behavior, unchanged) -------------------------
    if not use_kernel:
        proxy = parse_proxy_url(raw)
        if not proxy:
            return None
        auth = " (auth)" if proxy["username"] else ""
        return _build_helper(
            proxy, f"{proxy['kind']}://{proxy['hostname']}:{proxy['port']}{auth}")

    # --- kernel path (advanced protocols via xray/sing-box) ----------------
    kernel = create_kernel({
        "link": raw if is_kernel_scheme(raw) else None,
        "configPath": kernel_opts.get("configPath"),
        "kernel": kernel_opts.get("kernel"),
        "xrayBin": kernel_opts.get("xrayBin"),
        "singboxBin": kernel_opts.get("singboxBin"),
        "socksPort": kernel_opts.get("socksPort"),
    })
    socks_port = kernel.start()
    atexit.register(kernel.stop)
    proxy = {"kind": "socks5", "hostname": "127.0.0.1", "port": socks_port,
             "username": "", "password": ""}
    return _build_helper(proxy, f"kernel {kernel.describe()}", kernel=kernel)
